cub/cub/thread/thread_reduce.cuh

```diff
@@ -538,59 +411,53 @@ template <typename AccumT, typename Input, typename ReductionOp>
  * SIMD Reduction
  **********************************************************************************************************************/
 
-// never reached. Protect instantion of ThreadReduceSimd with arbitrary types and operators
-_CCCL_TEMPLATE(typename Input, typename ReductionOp)
-_CCCL_REQUIRES((!cub::internal::enable_generic_simd_reduction<Input, ReductionOp>()))
-[[nodiscard]] _CCCL_DEVICE _CCCL_FORCEINLINE auto ThreadReduceSimd(const Input& input, ReductionOp)
-  -> ::cuda::std::remove_cvref_t<decltype(input[0])>
+// NOTE: bit_cast cannot be always used because __half, __nv_bfloat16, etc. are not trivially copyable
+template <typename Output, typename Input>
+[[nodiscard]] _CCCL_DEVICE _CCCL_FORCEINLINE Output unsafe_bitcast(const Input& input)
 {
-  assert(false);
-  return input[0];
+  Output output;
+  static_assert(sizeof(input) == sizeof(output), "wrong size");
+  ::memcpy(static_cast<void*>(&output), static_cast<const void*>(&input), sizeof(input));
```

Comment:
Nit: The casts should not be necessary, since any `T*` should convert to `void*`
```suggestion
  ::memcpy(&output, &input, sizeof(input));
```